[OpenMP][Offload] Support standalone build

saiislam opened the pull request:
The standalone offload build will just search for
the ompt header and if found then OMPT support will
be enabled. This caused the LIBOMP_HAVE_VERSION_SCRIPT_FLAG
detection and pythonize_bool to move into offload/CMakeLists.txt.

If the cmake compiler is clang, get the resource directory for
installation of openmp headers. find_package(LLVM) is already
called in LibomptargetGetDependencies.

Depends on #75125

jdoerfert committed:
[Offload] Move `/openmp/libomptarget` to `/offload`

With this commit, users need to enable the new LLVM/Offload subproject
as a runtime in their cmake configuration. No further changes are
expected for downstream code.

Tests and other components still depend on OpenMP and have also not been
renamed. The results below are for a build in which OpenMP and Offload
are enabled runtimes. In addition to the pure git mv, I needed to adjust
some CMake files. Nothing is intended to change semantics.

```
ninja check-offload
```
Works with the X86 and AMDGPU offload tests

```
ninja check-openmp
```
Still works but doesn't build offload tests anymore.

```
ls install/lib
```
Shows all expected libraries, incl.
- `libomptarget.devicertl.a`
- `libomptarget-nvptx-sm_90.bc`
- `libomptarget.rtl.amdgpu.so` -> `libomptarget.rtl.amdgpu.so.19git`
- `libomptarget.so` -> `libomptarget.so.19git`

Fixes: https://github.com/llvm/llvm-project/issues/75124

jhuber6 commented on offload/CMakeLists.txt:
```diff
@@ -159,13 +168,20 @@ set (LIBOMPTARGET_TESTED_PLUGINS "")
 endif()
 
+# OMPD support for libomptarget (currently only with cuda)
+set(LIBOMPTARGET_OMPD_SUPPORT FALSE CACHE BOOL "OMPD-support?")
```
Why do we need `OMPD`?

jhuber6 commented on offload/CMakeLists.txt:
```diff
@@ -159,13 +168,20 @@ set (LIBOMPTARGET_TESTED_PLUGINS "")
 string( TOLOWER "${CMAKE_BUILD_TYPE}" LIBOMPTARGET_CMAKE_BUILD_TYPE)
 if(LIBOMPTARGET_CMAKE_BUILD_TYPE MATCHES debug)
   option(LIBOMPTARGET_ENABLE_DEBUG "Allow debug output with the environment variable LIBOMPTARGET_DEBUG=1" ON)
+  add_definitions(-DDEBUG)
```
What's this for? I don't think we want a global definition `DEBUG` set for every compilation.

jhuber6 commented on offload/CMakeLists.txt:
```diff
@@ -10,6 +10,13 @@
 
+cmake_minimum_required(VERSION 3.20.0)
+
+if (OPENMP_STANDALONE_BUILD OR "${CMAKE_SOURCE_DIR}" STREQUAL "${CMAKE_CURRENT_SOURCE_DIR}")
+  set(OPENMP_STANDALONE_BUILD TRUE)
+  project(omptarget C CXX ASM)
+endif()
```
Why "OpenMP?" Shouldn't it of `offload`? This is definitely required for standalone compilation however.

estewart08 replied:
Because the libomptarget folder was just renamed offload, all of the checks were not renamed. So we stuck with OPENMP_STANDALONE_BUILD.

jhuber6 replied:
I meant the `project` name as well.

What is this pattern
```cmake
if (OPENMP_STANDALONE_BUILD)
  set(OPENMP_STANDALONE_BUILD TRUE)
endif()
```
Seems wrong.

estewart08 replied:
Sure, the project could probably change. If the offload/CMakeLists.txt is invoked directly without setting `OPENMP_STANDALONE_BUILD` in the cmake configuration then `OPENMP_STANDALONE_BUILD` will need to be set here so the rest of the checks will work.

jhuber6 commented on cmake/Modules/OffloadOpenmpCommon.cmake:
```diff
@@ -0,0 +1,51 @@
+# Check if OMPT support is available
+# Currently, __builtin_frame_address() is required for OMPT
+# Weak attribute is required for Unices (except Darwin), LIBPSAPI is used for Windows
+if(NOT LIBOMP_ARCH)
```
Why do we need LIBOMP_ARCH here?

estewart08 replied:
Because right below LIBOMP_HAVE_OMPT_SUPPORT needs it and the standalone build of offload is not aware of LIBOMP_ARCH.

jhuber6 replied:
Isn't there some header that's only built if we have OMPT support? Presumably the presence of that would indicate that whoever built OpenMP had OMPT support, or we could just make it an option, there's no explicit need for this to be automatically detected since a standalone build is supposed to be manual.

estewart08 replied:
At first thought this could work. If the ompt header is found offload standalone would set the OMPT support to ON. This would allow us to drop the majority of this top level cmake module. Then that makes the argument for removing it entirely and just put `LIBOMP_HAVE_VERSION_SCRIPT_FLAG` and `pythonize_bool` directly into offload/CMakeLists.txt.

jhuber6 commented on offload/CMakeLists.txt:
```diff
@@ -190,33 +206,95 @@ if (LIBOMPTARGET_USE_LTO)
+endif()
+include_directories(${LIBOMPTARGET_INCLUDE_DIR} ${LIBOMP_OMP_TOOLS_INCLUDE_DIR})
+
+# Various LLVM_TOOLS are needed to build libomptarget
```
Where is this stuff coming from?

estewart08 replied:
Yeah this code below should not be necessary.

jhuber6 reviewed (commented):
I'm a little confused here. I thought all that's required is

1. Detect support for `--version-script` in the linker
2. Find `libomp.so` to link against it
3. Find `omp.h` and `omp-tools.h` to include.

What's all this other stuff for?

saiislam committed:
[Offload] Support standalone build for OpenMP and Offload

The standalone offload build will just search for
the ompt header and if found then OMPT support will
be enabled. This caused the LIBOMP_HAVE_VERSION_SCRIPT_FLAG
detection and pythonize_bool to move into offload/CMakeLists.txt.

If the cmake compiler is clang, get the resource directory for
installation of openmp headers. find_package(LLVM) is already
called in LibomptargetGetDependencies.

Depends on #75125

jhuber6 commented on offload/CMakeLists.txt:
```diff
@@ -35,11 +42,12 @@ endif()
 set(OFFLOAD_INSTALL_LIBDIR "lib${LLVM_LIBDIR_SUFFIX}")
 
-set(LLVM_COMMON_CMAKE_UTILS ${CMAKE_CURRENT_SOURCE_DIR}/cmake)
+set(LLVM_COMMON_CMAKE_UTILS ${CMAKE_CURRENT_SOURCE_DIR}/../cmake)
```
This doesn't make sense to me, the whole point of a standalone build is that it's separate from LLVM, which would mean that this is the top level directory in the tree. Why are we just assuming that the LLVM modules directory is there?

I don't think putting this module stuff is necessary since we can just make the users pass in `OMPTARGET_OMPT_ENABLED` or something.

estewart08 replied:
I think it's ok to assume the openmp sources are still inside the llvm-project, right? It is standalone in the sense that it will not be built during the same build as llvm. This change was more so a correction on the current offload patch `${CMAKE_CURRENT_SOURCE_DIR}/cmake` == `offload/cmake`, which is not what `LLVM_COMMON_CMAKE_UTILS` is supposed to reference.

I was just keeping the autodetection in for OMPT as the system configuration still needs checked to verify whether it can be supported, regardless if the user passes it in or not.

jhuber6 commented on offload/CMakeLists.txt:
```diff
@@ -190,6 +198,10 @@ if (LIBOMPTARGET_USE_LTO)
 endif()
 
+# Set LIBOMP_HAVE_OMPT_SUPPORT and LIBOMP_HAVE_VERSION_SCRIPT_FLAG using common offload/openmp cmake module.
+list(APPEND CMAKE_MODULE_PATH "${CMAKE_CURRENT_SOURCE_DIR}/../openmp/runtime/cmake")
+include(OffloadOpenmpCommon)
```
Again, we're assuming that this is in an LLVM directory despite the fact that this is supposed to be "standalone".

jhuber6 commented on offload/CMakeLists.txt:
```diff
@@ -214,6 +226,31 @@ set(LIBOMPTARGET_GPU_LIBC_SUPPORT ${LLVM_LIBC_GPU_BUILD} CACHE BOOL
+      omp-tools.h
+    HINTS
+    ${LLVM_INCLUDE_DIRS}
+    ${LLVM_INCLUDE_DIRS}/../lib/clang/${LLVM_VERSION_MAJOR}/include
```
`LLVM_VERSION_MAJOR` won't be set if this doesn't go through the LLVM CMake. You'll need to call `--print-resource-dir` if the compiler is clang.

estewart08 replied:
It will be set from the LLVMConfig.cmake from the result of find_package(LLVM). I did query `clang --print-resource-dir` in a version before this but thought this was a cleaner way.

jhuber6 commented on offload/CMakeLists.txt:
```diff
@@ -214,6 +226,31 @@ set(LIBOMPTARGET_GPU_LIBC_SUPPORT ${LLVM_LIBC_GPU_BUILD} CACHE BOOL
 pythonize_bool(LIBOMPTARGET_GPU_LIBC_SUPPORT)
 
+if(OPENMP_STANDALONE_BUILD)
+  if(CMAKE_CXX_COMPILER_ID MATCHES "Clang")
```
Don't we already do `find_package(LLVM)` somewhere else? And why only Clang.

jhuber6 commented on offload/plugins-nextgen/host/CMakeLists.txt:
```diff
@@ -33,7 +33,7 @@ endif()
 # Install plugin under the lib destination folder.
 install(TARGETS omptarget.rtl.${machine}
-        LIBRARY DESTINATION "${OPENMP_INSTALL_LIBDIR}")
+        LIBRARY DESTINATION "${OFFLOAD_INSTALL_LIBDIR}")
```
Why is this the only one that changed? In any case this will be completely gone if we land the static library patches.

estewart08 replied:
I think this should have been changed in the original offload patch, just caught it here.

jhuber6 commented on openmp/runtime/cmake/config-ix.cmake:
```diff
@@ -18,6 +18,7 @@ include(CheckIncludeFiles)
 include(CheckSymbolExists)
 include(LibompCheckFortranFlag)
 include(LLVMCheckCompilerLinkerFlag)
+include(OffloadOpenmpCommon)
```
I still don't think this is necessary. Just check the version script independently in `offload/` and make OMPT support opt-in.

saiislam commented:
Thanks @jhuber6 for the detailed review and @estewart08 for working on the patch.

@jdoerfert latest patch is much smaller and cleaner, and should be easy to squash with your patch.

jhuber6 reviewed (commented):
Looks better now, if it works for ROCm's purposes then let's ship it.

jdoerfert commented:
> Looks better now, if it works for ROCm's purposes then let's ship it.

@saiislam @estewart08 Does this work for ROCm? Can I pick it up?

jplehr commented:
The content of this PR has been pulled-in into the libomptarget->offload PR.
Closing this one.
